### PR TITLE
add additional functional api tests for Job Creation

### PR DESCRIPTION
**Pull request by mrdubr**
Additional functional API tests for the Job Creation flows per
https://pagerduty.atlassian.net/browse/RUN-2611

Note: Depends on https://github.com/rundeck/rundeck/pull/9283

### CODE REVIEW

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+        when:
+        def response = JobUtils.createJob(TEST_PROJECT, jobXml, client)
+
+        then:
+        response.successful
+        def jobId = MAPPER.readValue(response.body().string(), CreateJobResponse.class).getSucceeded().get(0).id
+        def completedJob = runJobAndWait(jobId)
+        // Should run the local node only since the filter excluded the nodes added by the setup
+        getOrderedNodesListExecutedOn(completedJob).size() == 1
```
Comment (smartinellibenedetti) [inline functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy:109]: same suggestion as the previous test

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobSpec.groovy [hunk 2]
```diff
@@ -1,32 +1,208 @@
...
+
+        when:
+        def response = JobUtils.createJob(PROJECT_NAME, jobXml, client)
+
+        then:
+        response.successful
+        def jr = MAPPER.readValue(response.body().string(), CreateJobResponse.class)
+        def jobDetails = JobUtils.getJobDetailsById(jr.getSucceeded().get(0).id, MAPPER, client)
+        jobDetails.scheduleEnabled == true
+    }
```
Comment (smartinellibenedetti) [inline functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobSpec.groovy:105]: question: It seems like there might be some tests in the JobExecutionSpec that also check if a job is being created with schedule/execution enabled/disabled (the tests starting with `test-job-flip-` name).

Would it be worth moving the tests added to JobSpec there and reorganizing the suite, or removing the tests added here as it seems that there's an overlap?

**Review by smartinellibenedetti — COMMENTED** [review]
Left a few suggestions and questions, but there are no major blockers.

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy (hunk 1 above)
Comment (mrdubr) [reply functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy:109]: The name of the local node is dynamic, thus I cannot do a simple list comparison.

File: functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/execution/CommandOnMultipleNodesSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public static final String TEST_SSH_ARCHIVE_DIR = "/projects-import/core-jsch-executor-test"
+    public static final String NODE_KEY_PASSPHRASE = "testpassphrase123"
+    public static final String NODE_USER_PASSWORD  = "testpassword123"
+    public static final String USER_VAULT_PASSWORD = "vault123"
+
+    @Override
+    void startEnvironment() {
+        String keyPath = getClass().getClassLoader().getResource("docker/compose/oss").getPath()+"/keys"
+
+        super.startEnvironment()
```
Comment (smartinellibenedetti) [inline functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/execution/CommandOnMultipleNodesSpec.groovy:22]:
```suggestion
```

If I recall correctly the super.starEnvironment isn't necessary

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+              <sequence>
+                <command>
+                  <exec>echo 0</exec>
+                </command>
+              </sequence>
+           </job>
+        </joblist>"""
+
+        when:
+        def response = JobUtils.createJob(TEST_PROJECT, jobXml, client)
```
[the remark this answers is not in the document]
Comment (mrdubr) [reply functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy:63]: I was not aware of these generators.
I can see how `updateJobFileToImport` has good intentions, but its implementation prevents meaningful reuse. Well... let me refactor it!

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobSpec.groovy (hunk 2 above)
Comment (mrdubr) [reply functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobSpec.groovy]: Thanks for pointing this out. I was able to make a few minor changes to the tests in `JobExecutionSpec` and eliminate a few from this suite.

File: functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrdubr) [reply functional-test/src/test/groovy/org/rundeck/tests/functional/api/job/JobExecutionOnMultipleNodesSpec.groovy]: Whew... :-) 
I had to do waaay more refactoring than I was planning to due to the `Job` deserialization bug, but pretty happy with the result as I was able to eliminate XML from the tests.

File: functional-test/src/test/groovy/org/rundeck/util/common/jobs/JobUtils.groovy [hunk 5]
```diff
@@ -252,32 +253,46 @@ class JobUtils {
...
      *                 - "uuid": UUID for the job (default: generated UUID)
+     *                 - "execution-enabled": Execution enabled flag (default: "true")
+     *                 - "schedule-enabled": Schedule enabled flag (default: "true")
+     *                 - "node-filter-include": Node filter include (default: ".*")
+     *                 - "node-filter-exclude": Node filter exclude (default: "")
+     *                 - "dispatch-rank-order": Dispatch rank order (default: "ascending")
      * @return The path of the updated temporary XML file.
      */
-    static def updateJobFileToImport = (String fileName, String projectName, Map args = [:]) -> {
-        if (args.isEmpty()) {
```
Comment (mrdubr) [inline functional-test/src/test/groovy/org/rundeck/util/common/jobs/JobUtils.groovy:270]: This was unsustainable. Fortunately, the refactoring appears to be backward compatible.

File: functional-test/src/test/groovy/org/rundeck/util/common/jobs/JobUtils.groovy [hunk 6]
```diff
@@ -112,9 +113,9 @@ class JobUtils {
                 "</joblist>"
     }
 
-    static Job getJobDetailsById(final String jobId, final ObjectMapper mapper, RdClient client){
+    static JobDetails getJobDetailsById(final String jobId, final ObjectMapper mapper, RdClient client){
```
Comment (mrdubr) [inline functional-test/src/test/groovy/org/rundeck/util/common/jobs/JobUtils.groovy:116]: Deserialization was failing when the job had the "options" property. 
`Job` expected the `Map`, but a list of object was returned from the API. Had to introduce `JobDetails`

**Review by smartinellibenedetti — APPROVED** [review]
PR is looking great, thank you for the refactoring!

**Pull request merged**